### DISCUSSION TITLE
A better way to do stabilization

### DISCUSSION THREAD

**Discussion by reverendbedford** [idaholab/moose#17569] [Q&A General] [answered]
**Stabilization**

A lot of different physics problems require stabilization.  The example I'm most familiar with is incompressible deformation in solid mechanics where you need to alter the finite element space to produce non-pathological results for some problems.  The classical way to do this for hexes is B_bar or F_bar (small/large deformations) methods, where you average the volumetric components of the shape function gradients over each element.  But plenty of other physics have similar problems and solutions, for example stabilization advection equations like in the Navier Stokes module through upwinding, Petrov-Galerkin, GLS, etc.

**How we do stabilization now**

Currently MOOSE users tend to do this in Kernel/Material land.  For example, you can implement B_bar methods by averaging the volumetric strain over the element, using that strain in the stress update, and adding a bunch of terms to the kernel Jacobian to account for the averaging.

I would argue that's the "old" way to think of things (i.e. changing the material kinematics) and produces not-very-clean code.  For example, you might have more than one kernel implementing similar physics (advection-diffusion-reaction and Navier-Stokes or the new total and updated Lagrangian kernels I've proposed for tensor mechanics, for example) and the variants could be stabilized with the same scheme.  But if you're doing it in materials/kernels you would have to write it twice.  It also tends to lead towards ugly code.  For example, in B_bar the stabilization actually happens in the strain calculator but you need to know what happened, i.e. what type of stabilization, in the kernel.

**A better way to do it**

A better way to think of stabilization might be as altering the underlying finite element space, i.e. the test and trial functions themselves.  A more "natural" method for doing stabilization would then be some class (a subclass of variable, I guess) that alters the test and trial functions and their gradients coming from libmesh to provide the stabilization.  This type of implementation would greatly simplify things on the material/kernel side -- you would use exactly the same implementation for stabilized and unstabilized versions, you wouldn't need to "know" about stabilization at all -- and make it much easier to add/try out new stabilization techniques.

Unfortunately my MOOSE skills aren't developed enough to know how to do this/how easy this is/if it's even possible.  At a minimum this class would need to:

- Be interchangeable with the existing Variable system (i.e. provide variable values and gradients and test/trial function values and derivatives).
- Have access to more than one variable -- for example, B_bar stabilization would need to know about all three displacement components in a solids kernel (i.e. disp_x, disp_y, disp_z).
- Take constant stabilization parameters.

A more sophisticated version would do these things plus:

- Access constant or lagged material properties to set appropriate values for stabilization parameters.
- For the most modern methods use some weird FE spaces, often involving bubble functions.  This isn't my area, others may know more.  The advantage is typically they can provide "optimal" stabilization without any tuning parameters.

_At this point I'm just looking for comments on:_

- If this is a good idea
- If it's possible
- Would it require libmesh modifications (I don't think so...)

**Comment by WilkAndy** [comment]
This is an interesting thought, @reverendbedford .   My unthinking response is: is it possilbe do use this approach for advection stabilization, such as full upwinding or KT or RDG, eg see my explanation of KT at https://mooseframework.inl.gov/modules/porous_flow/kt_worked.html ?  I don't see exactly how your idea works.  Would evaluating your test functions involve assessing advective flows in neighbouring elements?

**Comment by jiangwen84** [comment]
@reverendbedford do you have any ideas on how to modify the finite element space for appropriate stabilization? I guess the modification might depend on your materials, i.e. volumetric locking, shearing locking etc.

**Reply by lindsayad** [reply]
I agree with @jiangwen84 here. Almost always, it seems to me, the method of stabilization is dependent on the physics.

**Comment by lindsayad** [comment, marked as answer]
It doesn't seem overly natural to me to have this done at the `Variable` level although maybe I could be convinced.

A leading obstacle to me to some generic implementation is that a classic Petrov-Galerkin implementation for, say, stabilizing CFEM Navier-Stokes equations has a form like:

test_star = test + tau * grad_test * velocity

However, this doesn't really apply for terms that are integrated by parts, e.g. the viscous term or even the pressure term. So then one still has to be aware of the physics that they are doing. I just think it's hard to fully automate. You mentioned a stabilization for solid mechanics...that's clearly different from the stabilization I posted above with a velocity. To generalize it perhaps you could create some interface that takes a lambda that would implement the stabilization you want...but then again you are requiring *somebody* to do some thinking.

FWIW we do have some general objects that are meant to help application developers with stabilization. But these are still kernels. They are `ADKernelStabilized.[hC]` and `ADKernelSUPG.[hC]`, the latter of which implements a stabilization method which is virtual in `ADKernelStabilized`. We use these in the `navier_stokes` module for CFEM.

**Comment by reverendbedford** [comment]
I agree that the stabilization will still ultimately depend on the physics, but I guess my argument is that multiple implementations of similar physics might want to use the same stabilization scheme.  So for example I have my updated and total Lagrangian solids kernels and both could use the same stabilization even though the weak residuals look very different.

That said, my initial argument was Petrov-Galerkin where you can alter the test functions and not touch anything on the material/kernel side.  But I see your point, you still need to know when/how to apply it in kernel-space.

Interestingly the classical way to stabilize solids is to alter the *trial* functions, not the test functions.  This may mean it propagates through the formulation easier.  But it also may mean this would be too specialized to be worth the effort.  Either way, let me do some math, I'll post the results here.

**Reply by hugary1995** [reply]
I think some equations can greatly help here. I get that your approach would benefit stabilization implementations in MOOSE in general, but it may be easier if we first limit our discussion to a very specific stabilization. I guess we are mostly interested in mechanics for now, so how about just a few equations on how you want to B-bar? If we find a path forward we can then talk about advection stabilizations.
